web-api/src/personalization/filter.rs

```diff
@@ -444,227 +519,114 @@ mod tests {
             },
         );
-        assert_eq!(
-            serde_json::from_str::<CompareWith>("{}")
-                .unwrap_err()
-                .to_string(),
-            format!(
-                "invalid length 0, expected {} at line 1 column 2",
-                CompareWith::EXPECTING,
-            ),
+        assert_invalid_len::<CompareWith>("{}", 0, 2);
+        assert_invalid_len::<CompareWith>(
+            r#"{ "$in": ["a", "b", "c"], "$in": ["d", "e", "f"] }"#,
+            2,
+            50,
         );
-        assert_eq!(
-            serde_json::from_str::<CompareWith>(
-                r#"{ "$in": ["test", "this"], "$in": ["test", "that"] }"#
-            )
-            .unwrap_err()
-            .to_string(),
-            format!(
-                "invalid length 2, expected {} at line 1 column 52",
-                CompareWith::EXPECTING,
-            ),
-        );
-        assert_eq!(
-            serde_json::from_str::<CompareWith>(r#"{ "$in": "test" }"#)
-                .unwrap_err()
-                .to_string(),
-            format!(
-                "invalid type: string \"test\", expected {} at line 1 column 17",
-                CompareWith::EXPECTING,
-            ),
-        );
-
-        let array = vec!["test"; CompareOp::MAX_VALUES_PER_IN + 1];
-        assert_eq!(
-            serde_json::from_str::<CompareWith>(&json!({ "$in": array }).to_string())
-                .unwrap_err()
-                .to_string(),
-            format!(
-                "invalid length {}, expected {} at line 1 column 3516",
-                CompareOp::MAX_VALUES_PER_IN + 1,
-                CompareWith::EXPECTING,
-            ),
+        assert_invalid_len::<CompareWith>(
+            &json!({ "$in": vec!["a"; CompareOp::MAX_VALUES_PER_IN + 1] }).to_string(),
+            CompareOp::MAX_VALUES_PER_IN + 1,
+            2013,
         );
+        assert_invalid_type(r#"{ "$eq": ["a", "b", "c"] }"#, "sequence", 26);
+        assert_invalid_type(r#"{ "$gt": ["a", "b", "c"] }"#, "sequence", 26);
+        assert_invalid_type(r#"{ "$gte": ["a", "b", "c"] }"#, "sequence", 27);
+        assert_invalid_type(r#"{ "$lt": ["a", "b", "c"] }"#, "sequence", 26);
+        assert_invalid_type(r#"{ "$lte": ["a", "b", "c"] }"#, "sequence", 27);
     }
 
     #[test]
     fn test_compare_with_number() {
-        for operation in [
+        for (op, operation) in [
             ("$gt", CompareOp::Gt),
             ("$gte", CompareOp::Gte),
             ("$lt", CompareOp::Lt),
             ("$lte", CompareOp::Lte),
         ] {
             for number in [json!(42_u64), json!(-42_i64), json!(42_f64)] {
                 assert_eq!(
-                    serde_json::from_str::<CompareWith>(
-                        &json!({ operation.0: number }).to_string()
-                    )
-                    .unwrap(),
+                    serde_json::from_str::<CompareWith>(&format!(r#"{{ "{op}": {number} }}"#))
```

Comment:
:wrench:  Why not just use `json`? 

(here and below where you sometimes explicitly changed away from it).

Reply:
i'm a bit undecided about this, in the end the outcome is of course the same if we use `format!()` or `json!().to_string()`. currently i'm leaning more towards the former as this seems a bit closer to what we would get in a request written by a user. but i also the that the additional escaping in the formatting is not nice. should we go back to the latter, wdyt?

Reply:
I think  `json!` is more readable and guarantees that the json itself is always well formed and as such should be preferred in this context as we are not testing the json parsing itself.

Any performance differences are in this use-case irrelevant AFIK.